File: backend/app/schemas/device.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from app.models.enum.device_connectivity import DeviceConnectivity
    
class DeviceCreate(BaseModel):
    uid: str
    name: str
    department_id: int
    data_interval: float
    machine: Optional[str] = None
    connectivity: DeviceConnectivity = DeviceConnectivity.wifi
    mobile_number: Optional[str] = None
    sim_id: Optional[str] = None

class DeviceUpdate(BaseModel):
    name: Optional[str] = None
    machine: Optional[str] = None
    department_id: Optional[int] = None
    data_interval: Optional[float] = None
    dashboard_config: Optional[dict] = None
    is_online: Optional[bool] = None
    is_active: Optional[bool] = None
    connectivity: Optional[DeviceConnectivity] = None
    mobile_number: Optional[str] = None
    sim_id: Optional[str] = None


class DeviceOut(BaseModel):
    id: int
    uid: str
    name: str
    machine: Optional[str] = None
    connectivity: DeviceConnectivity
    mobile_number: Optional[str] = None
    sim_id: Optional[str] = None
    data_interval: float
    dashboard_config: Optional[dict] = None
    department_name: str
    customer_name: str
    customer_mqtt_topic: str
    distributor_name: Optional[str] = None
    created_at: datetime
    is_online: bool
    is_active: bool


    class Config:
        from_attributes = True

class DeviceListResponse(BaseModel):
    items: List[DeviceOut]
    total: int
    page: int
    page_size: int

class DeviceRecentOut(BaseModel):
    id: int
    uid: str
    name: str
    department_name: str
    customer_name: str
    customer_mqtt_topic: str
    is_online: bool

    class Config:
        from_attributes = True

File: frontend/IoT_Monitoring_System/src/types/device.ts
export type DataFieldType = "number" | "text" | "list" | "boolean";

export type GridLayoutItem = {
  i: string;
  x: number;
  y: number;
  w: number;
  h: number;
  minW?: number;
  minH?: number;
};

export type DashboardSection = {
  id: string;
  name: string;
  collapsed?: boolean;
};

export type DataPanelFieldConfig = {
  label?: string;
  unit?: string;
  type?: DataFieldType;
  color?: string;
  cases?: string[] | null;
  case_colors?: Record<string, string> | null;
  true_label?: string;
  false_label?: string;
  true_color?: string;
  false_color?: string;
  section_id?: string | null;
};

export type DashboardPanelConfig = {
  fields?: string[];
  config?: Record<string, DataPanelFieldConfig>;
  layout?: GridLayoutItem[];
  sections?: DashboardSection[];
};

export type DashboardChartItem = {
  id: string;
  type: "meter" | "line" | "area" | "pie" | "stat" | "bar" | "button";
  field: string;
  name?: string | null;
  min?: number | null;
  max?: number | null;
  tick_count?: number | null;
  value_decimals?: number | null;
  stat_font_size?: number | null;
  value_cases?: string[] | null;
  fields?: string[] | null;
  section_id?: string | null;
  bar_orientation?: "horizontal" | "vertical" | null;
  bar_race_mode?: boolean | null;
  pie_show_labels?: boolean | null;
  line_list_mode?: "single" | "multi" | null;
  output_value_type?: "boolean" | "multi" | null;
};

export type DashboardChartConfig = {
  items?: DashboardChartItem[];
  layout?: GridLayoutItem[];
  sections?: DashboardSection[];
};

export type DashboardConfig = {
  data_panel?: DashboardPanelConfig;
  data_chart?: DashboardChartConfig;
};

export type LegacyDashboardConfig = {
  data_panel_fields?: string[];
  data_panel_config?: Record<string, DataPanelFieldConfig>;
  data_panel_layout?: GridLayoutItem[];
  data_panel_sections?: DashboardSection[];
  data_chart_items?: DashboardChartItem[];
  data_chart_layout?: GridLayoutItem[];
  data_chart_sections?: DashboardSection[];
};

export type DeviceDashboardConfig = DashboardConfig | LegacyDashboardConfig;

export type DeviceConnectivity = "wifi" | "cellular";

export interface Device {
  id: number;
  name: string;
  uid: string;
  machine?: string | null;
  connectivity: DeviceConnectivity;
  mobile_number?: string | null;
  sim_id?: string | null;
  sub?: string | null;
  pub?: string | null;
  data_interval: number;
  dashboard_config?: DeviceDashboardConfig | null;
  is_online: boolean;
  department_name?: string | null;
  customer_name?: string | null;
  is_active: boolean;
  created_at: string;
}

export interface DeviceListResponse {
  items: Device[];
  total: number;
  page: number;
  page_size: number;
}
